javatests/arcs/core/analysis/ExpressionClaimDeducerTest.kt

```diff
@@ -0,0 +1,381 @@
+    val expr = PaxelParser.parse("new Object {foo: x, bar: y}")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Scope(
+        "foo" to Deduction.Equal(Deduction.Path("x")),
+        "bar" to Deduction.Equal(Deduction.Path("y"))
+      )
+    )
+  }
+
+  @Test
+  fun new_field_access() {
+    val expr = PaxelParser.parse("new Object {foo: input.foo, bar: input.foo.bar}")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Scope(
+        "foo" to Deduction.Equal(Deduction.Path("input", "foo")),
+        "bar" to Deduction.Equal(Deduction.Path("input", "foo", "bar"))
+      )
+    )
+  }
+
+  @Test
+  fun sum_variables() {
+    val expr = PaxelParser.parse("x + y")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Derive(
+        Deduction.Paths(
+          listOf("x"),
+          listOf("y")
+        )
+      )
+    )
+  }
+
+  @Test
+  fun new_field_access_binexpr() {
+    val expr = PaxelParser.parse("new Object {foo: input.foo, bar: input.foo.bar + input.foo}")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Scope(
+        "foo" to Deduction.Equal(Deduction.Path("input", "foo")),
+        "bar" to Deduction.Derive(
+          Deduction.Paths(
+            listOf("input", "foo", "bar"),
+            listOf("input", "foo")
+          )
+        )
+      )
+    )
+  }
+
+  @Test
+  fun field_access_nested_expr() {
+    val expr = PaxelParser.parse("(new Foo {x: foo, y: bar}).x")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Scope(
+        "x" to Deduction.Equal(Deduction.Path("foo"))
+      )
+    )
+  }
+
+  @Test
+  fun field_access_nested_multi_field_expr() {
+    val expr = PaxelParser.parse("(new Foo {a: foo.x + foo.y, b: foo.y + foo.z}).a")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Scope(
+        "a" to Deduction.Derive(
+          Deduction.Paths(
+            listOf("foo", "x"),
+            listOf("foo", "y")
+          )
+        )
+      )
+    )
+  }
+
+  @Test
+  fun new_nested() {
+    val expr = PaxelParser.parse(
+      """
+      new Foo {
+        a: new Bar {
+          x: cat,
+          y: dog
+        },
+        b: foo,
+        c: 5
+      }
+      """.trimIndent())
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Scope(
+        "a" to Deduction.Scope(
+          "x" to Deduction.Equal(Deduction.Path("cat")),
+          "y" to Deduction.Equal(Deduction.Path("dog"))
+        ),
+        "b" to Deduction.Equal(Deduction.Path("foo")),
+        "c" to Deduction.Empty
+      )
+    )
+  }
+
+  @Test
+  fun new_deeply_nested() {
```

Comment:
Another test for you to check alias substitutions with deep structures:
```
from f in foo
let x = new Foo {
  y: new Foo {
    z: f.a
  }
}
select x.y.z
```

It should produce Deduction.Equal(foo.a)

Reply:
I don't have let implemented yet, however, the rest of this example would make a great test case.